server/EXCEPTIONS.md

```diff
@@ -0,0 +1,168 @@
+
+```js
+// services/my-project.service.ts
+getProject(){
+  try {
+    return myProject = this.get('dcp_project', ...);
+  } catch (e) {
+    throw new HttpException({
+      code: 'GET_PROJECTS',
+      message: `Failed to find a project for given id. CRM: ${e}`,
+    });
+  }
+}
+```
+
+Generally, when generating errors as described above, <u>*do so within Nest services*</u>.
+
+## Relaying Errors
+
+In the controller that consumes a service, _relay_ the error from the service by simply throwing the error as it bubbles up:
+
+```js
+// controllers/my-project.controller.ts
+@Get('/')
+  async getProject(@Session() session) {
+    try {
+      return this.myProjectService.getProject();
+    } catch (e) {
+      throw e; // simply relay the bubbling error
+    }
+  }
+}
+```
+
+Note how it is not necessary to construct a new HttpException in the controller to relay the error from the service. Instead, we simply use `throw` to pass along the error `e` from the service.
+
+Errors will *bubble* through the function call stack. The `catch` block in a try-catch will also catch errors generated from *within* functions that are called in the `try` block.
+
+Only the HttpException caught and thrown closest to or on the controller
+endpoint will be returned for a given request. To prevent confusion over which
+backend layer should provide the most detailed error handling, this document prescribes generating new HTTPExceptions in services, and simply relaying them in controllers.
+
+## General Errors / Preventing Gaps
+Since we are trying to `throw new HttpException` at the service level, we may
+want to write a generic, controller-level catch-all for errors. This will help catch
+any uncaught errors from our services. Services are where we house the bulk
+of our business logic for processing requests, so they have a large surface
+area over which an error may not yet be covered by (occur outside of) a
+try-catch.
+
+```js
+// controllers/my-project.controller.ts
+@Get('/')
+async getProject(@Session() session) {
+  try {
+    return this.myProjectService.getProject();
+  } catch (e) {
+    if (e instanceof HttpException) {
+      throw e;
+    else {
+      // A catch-all, providing a general error
+      throw new HttpException({
+        code: 'GET_PROJECT_UNKNOWN',
+        message: 'An unknown server error occured while getting projects',
+      }, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+}
+```
+
+## HTTP Exception Filter
+
+Our nest backend has a custom [Nest exception filter](https://docs.nestjs.com/exception-filters#throwing-standard-exceptions) in place to catch all thrown HttpExceptions and format them before relaying them to the frontend. It will wrap the errors inside an `errors` array in an object:
+
+```json
+  {
+    errors: [ ...thrownHttpExceptions]
+  }
+```
+
+This means in our code we don't need to manually wrap the error. Below is what manually wrapping the error would look like, but our HttpException does this for us:
```

Comment:
You show an example below of our code if we were handling it manually, can you also show an example of the code where we don't handle it manually (so how we have it now since HttpException handles it for us)? Would be nice to see explicit comparison side by side

Reply:
The contrasting example would actually be the one in the previous section, "General Errors / Preventing Gaps". I'll clarify this!

Reply:
okay sweet, thanks!